### PR TITLE
Add descriptive docstring to WhisperTimeStampLogitsProcessor

### LINKED ISSUE #24783: Generate: have an example on each `LogitsProcessor` class docstring
# Context

`.generate()` can be extensively manipulated through `LogitsProcessor` (and `LogitsWarper`) classes. Those classes are the code implementation behind flags like `temperature` or `top_k`. 

Most of our `LogitsProcessor` classes have a docstring that briefly describes their effect. However, unless you are an expert in text generation, it's hard to fully grasp the impact of using each class. In some cases, it is also non-trivial to prepare the arguments to initialize the `LogitsProcessor` class. As such, each class should have a clear usage example with `.generate()` in their docstring! 💪 

Here is an example: [SequenceBiasLogitsProcessor docstring](https://github.com/huggingface/transformers/blob/f1732e1374a082bf8e43bd0e4aa8a2da21a32a21/src/transformers/generation/logits_process.py#L559). Contrarily to the other classes (at the time of writing), we can quickly learn how to use it just by reading its docstring. We are also immediately aware of a few caveats 🤓 

Bonus points: our docstring examples are part of our CI, so we would be beefing up our tests to ensure we don't add regressions 🤗 

This issue is part of the [text generation docs rework](https://github.com/huggingface/transformers/issues/24575).

# How to participate?
1. Ensure you've read our contributing [guidelines](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md) 📜 
2. Claim your `LogitProcessor` class in this thread (confirm no one is working on it). You can check the full list of classes below, and you can find their implementation in [this file](https://github.com/huggingface/transformers/blob/main/src/transformers/generation/logits_process.py) 🎯 
    - You may need to do some detective work to fully understand the purpose of the class. For instance, some classes were created as part of a paper to be applied to any model, others are model-specific, and some exist to avoid weird bugs 🕵️ 
    - Looking at the git history is a great way to understand how a `LogitsProcessor` came to be.
3. Implement your changes, taking the [SequenceBiasLogitsProcessor docstring](https://github.com/huggingface/transformers/blob/f1732e1374a082bf8e43bd0e4aa8a2da21a32a21/src/transformers/generation/logits_process.py#L559) as reference  💪 
    - Add a clear example that calls the processor through `.generate()`. Make sure the example's outputs are correct and that the model used in the test is a small model (anything larger than GPT2 needs explicit approval);
    - If you feel like the original docstring could be better, feel free to enhance it as well!
    - Don't forget to run `make fixup` before your final commit.
4. Open the PR and tag me in it  🎊 

# Tracker
- [x] MinNewTokensLengthLogitsProcessor
- [x] TemperatureLogitsWarper
- [x] RepetitionPenaltyLogitsProcessor
- [ ] EncoderRepetitionPenaltyLogitsProcessor
- [x] TopPLogitsWarper
- [ ] TopKLogitsWarper
- [ ] TypicalLogitsWarper
- [x] EpsilonLogitsWarper
- [x] EtaLogitsWarper
- [x] NoRepeatNGramLogitsProcessor
- [ ] EncoderNoRepeatNGramLogitsProcessor
- [x] SequenceBiasLogitsProcessor
- [x] NoBadWordsLogitsProcessor
- [ ] PrefixConstrainedLogitsProcessor
- [x] HammingDiversityLogitsProcessor
- [ ] ForcedBOSTokenLogitsProcessor
- [ ] ForcedEOSTokenLogitsProcessor
- [ ] InfNanRemoveLogitsProcessor
- [ ] ExponentialDecayLengthPenalty
- [ ] LogitNormalization
- [ ] SuppressTokensAtBeginLogitsProcessor
- [ ] SuppressTokensLogitsProcessor
- [ ] ForceTokensLogitsProcessor
- [ ] WhisperTimeStampLogitsProcessor
- [ ] ClassifierFreeGuidanceLogitsProcessor

### PR DESCRIPTION
**Pull request by jprivera44**
# What does this PR do?

This PR adds in docstrings that explains the usage of the arguments that can be passed in to the Whisper logits processor.

Fixes #24783 

## Before submitting
- [ x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).

## Who can review?
@gante

### CODE REVIEW

**Comment by ArthurZucker** [comment]
Hey! Thanks for opening a PR! 
This looks a bit too specific / long so will not be accepting it. Also you should run `pytest --doctest src/transformers/generation/logits_process.py` / rebase to main to check if this the docstring examples actually work 😉

File: src/transformers/generation/logits_process.py [hunk 1]
```diff
@@ -1330,8 +1330,14 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
 
 class WhisperTimeStampLogitsProcessor(LogitsProcessor):
     r"""
-    Whisper specific Processor. This processor can be used to force a list of tokens. The processor will set their log
-    probs to `inf` so that they are sampled at their corresponding index.
+    [LogitsProcessor]Whisper specific Processor. This processor is specifically designed for the Whisper Automatic
+    Speech Recognition model. It facilitates the manipulation of log probabilities for a predefined list of tokens
+    during text generation. By using this processor, you can effectively "force" certain tokens to be selected at
+    specific positions in the generated sequence. When tokens are passed to this processor, their log probabilities are
+    set to `inf` (infinity),ensuring that they are chosen at their corresponding indices. This is particularly useful
```
Comment (gante) [inline src/transformers/generation/logits_process.py:1337]:
```suggestion
    set to `inf` (infinity), ensuring that they are chosen at their corresponding indices. This is particularly useful
```

File: src/transformers/generation/logits_process.py [hunk 2]
```diff
@@ -1345,6 +1351,72 @@ class WhisperTimeStampLogitsProcessor(LogitsProcessor):
...
                     predicting timestamps that are too far in the future.
+
+
+    <Tip>
+
+    Make sure you set the sampling_rate` argument to this function below for the whisper model
+
+    </Tip>
+
+
```
Comment (gante) [inline src/transformers/generation/logits_process.py:1362]:
```suggestion
```
This would be a tip for the model, and not for the logits processor

File: src/transformers/generation/logits_process.py [hunk 3]
```diff
@@ -1345,6 +1351,72 @@ class WhisperTimeStampLogitsProcessor(LogitsProcessor):
...
+    ``` py
+    >>> import torch
+    >>> from transformers import AutoProcessor, WhisperForConditionalGeneration,GenerationConfig
+    >>> from datasets import load_dataset
+
+    >>> processor = AutoProcessor.from_pretrained("openai/whisper-tiny.en")
+    >>> model = WhisperForConditionalGeneration.from_pretrained("openai/whisper-tiny.en")
+
+    >>> generation_config = GenerationConfig.from_pretrained("openai/whisper-tiny.en")
+    >>> model.generation_config = generation_config
```
Comment (gante) [inline src/transformers/generation/logits_process.py:1373]:
```suggestion
```
This happens automatically under the hood, no need to hardcode it :)

File: src/transformers/generation/logits_process.py [hunk 4]
```diff
@@ -1345,6 +1351,72 @@ class WhisperTimeStampLogitsProcessor(LogitsProcessor):
...
+
+    >>> #Downloading an example audio file to process
+    >>> ds = load_dataset("hf-internal-testing/librispeech_asr_dummy", "clean", split="validation")
+    >>> file_index =4
+
+
```
Comment (gante) [inline src/transformers/generation/logits_process.py:1379]: Formatting: we only leave at most one line between example lines

**Review by gante — COMMENTED** [review]
Hi @jprivera44 👋 Thank you for opening the PR :) 

I have an important request here -- let's aim at having a short example with short outputs. ~20 lines of code is the sweet spot. Showing an example where we control returning the timestamps would be enough :)

**Comment by jprivera44** [comment]
Hi @gante thank you for your feedback :) I've made the suggested changes. However, I'm a little bit confused on the returning of the timestamps? The WhisperTimeStampLogitsProcessor only returns the scores, which then get transcribed into words. Would outputting the token timestamps be sufficient? Since these are directly accessible within modeling_whisper.py.

@ArthurZucker , I ran pytest --doctest-modules and all tests within the WhisperLogits passed. I noticed the recommendation for --doctest – does the transformers repo use a custom configuration that requires this argument?

**Comment by gante** [comment]
@jprivera44 Since the last reviews, we've added this file to the list of files to be doctested in our PR CI. If you rebase with `main`, your examples will be tested :) 

Regarding the examples themselves... something is odd. We should be getting a timestamp at the start of the transcription, right @ArthurZucker?

File: src/transformers/generation/logits_process.py [hunk 5]
```diff
@@ -1345,6 +1348,45 @@ class WhisperTimeStampLogitsProcessor(LogitsProcessor):
...
+    >>> model = WhisperForConditionalGeneration.from_pretrained("openai/whisper-tiny.en")
+    >>> ds = load_dataset("hf-internal-testing/librispeech_asr_dummy", "clean", split="validation")
+    >>> inputs = processor(ds[4]["audio"]["array"], return_tensors="pt")
+    >>> input_features = inputs.input_features
+
+    >>> #Change EOS:
+    >>> #This allows the user to select a specific token to terminate the sequence on, in this case it's the word poem(21247)
+    >>> model.generation_config.eos_token_id = 21247
+    >>> generated_ids = model.generate(inputs=input_features,return_timestamps=True)
+    >>> transcription = processor.batch_decode(generated_ids, skip_special_tokens=True)[0]
```
Comment (ArthurZucker) [inline src/transformers/generation/logits_process.py:1368]:
```suggestion
    >>> transcription = processor.batch_decode(generated_ids, skip_special_tokens=False)[0]
```
Let's have one example with timestamps no?

**Review by ArthurZucker — COMMENTED** [review]
The timestamps are ignored by the `skip_special_tokens = True`, let's try to show them

File: src/transformers/generation/logits_process.py (hunk 5 above)
Comment (gante) [reply src/transformers/generation/logits_process.py:1368]: TIL 👍

Comment (gante) [reply src/transformers/generation/logits_process.py:1368]: @jprivera44 let's a) limit the number of examples (one with and another without timestamps is enough); b) use Arthur's suggestion here :)

**Comment by jprivera44** [comment]
Hello @gante, apologies I was having some issues getting the time stamps to show up, but it's all fixed now. I've added in the updated code we discussed with two examples, one with timestamps and one without. This includes running pytest as @ArthurZucker mentioned. Please let me know what else is needed, and thank you again for the help!

**Comment by jprivera44** [comment]
Hello @gante hope you're well, is there anything else needed on this from my end? I just want to make sure I've completed the requirements.

File: src/transformers/generation/logits_process.py [hunk 6]
```diff
@@ -1345,6 +1348,34 @@ class WhisperTimeStampLogitsProcessor(LogitsProcessor):
...
+    >>> from transformers import AutoProcessor, WhisperForConditionalGeneration,GenerationConfig
+    >>> from datasets import load_dataset
+
+    >>> processor = AutoProcessor.from_pretrained("openai/whisper-tiny.en")
+    >>> model = WhisperForConditionalGeneration.from_pretrained("openai/whisper-tiny.en")
+    >>> ds = load_dataset("hf-internal-testing/librispeech_asr_dummy", "clean", split="validation")
+    >>> inputs = processor(ds[3]["audio"]["array"], return_tensors="pt")
+    >>> input_features = inputs.input_features
+
+    >>> #Displaying time time stamps
```
Comment (ArthurZucker) [inline src/transformers/generation/logits_process.py:1364]:
```suggestion
    >>> #Displaying timestamps
```

File: src/transformers/generation/logits_process.py [hunk 7]
```diff
@@ -1345,6 +1348,34 @@ class WhisperTimeStampLogitsProcessor(LogitsProcessor):
...
+    >>> generated_ids = model.generate(inputs=input_features, return_timestamps=True)
+    >>> transcription = processor.batch_decode(generated_ids, skip_special_tokens=False)[0]
+    >>> print("Transcription:", transcription)
+    Transcription: <|startoftranscript|><|0.00|> He has grave doubts whether Sir Frederick Layton's work is really Greek after all, and can<|6.44|><|6.44|> discover in it but little of rocky Ithaca.<|9.44|><|endoftext|>
+
+
+    >>> #No timestamps & change EOS:
+    >>> #This allows the user to select a specific token to terminate the sequence on, in this case it's the word poem(21247)
```
Comment (ArthurZucker) [inline src/transformers/generation/logits_process.py:1372]: looks like it was not stopped on poem no?

**Review by ArthurZucker — COMMENTED** [review]
Hey, @gante is OOO for a while; I'll take this over

File: src/transformers/generation/logits_process.py (hunk 7 above)
Comment (jprivera44) [reply src/transformers/generation/logits_process.py:1372]: Yes, the phrase stops on token in an order to showcase how changing the EOS_token parameter changes the output of the transcription. If preferred, I can only showcase the generating of timestamps vs. non-timestamps.

File: src/transformers/generation/logits_process.py [hunk 8]
```diff
@@ -1330,8 +1330,11 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
 
 class WhisperTimeStampLogitsProcessor(LogitsProcessor):
     r"""
-    Whisper specific Processor. This processor can be used to force a list of tokens. The processor will set their log
-    probs to `inf` so that they are sampled at their corresponding index.
+    [`LogitsProcessor`] Whisper specific Processor. This processor is specifically designed for the Whisper Automatic
+    Speech Recognition model. It facilitates the manipulation of log probabilities for a predefined list of tokens
+    during text generation. By using this processor, you can effectively "force" certain tokens to be selected at
+    specific positions in the generated sequence. When tokens are passed to this processor, their log probabilities are
+    set to `inf` (infinity), ensuring that they are chosen at their corresponding indices.
```
[the remark this answers is not in the document]
Comment (jprivera44) [reply src/transformers/generation/logits_process.py:1337]: Ah I see what you mean, I'll go ahead and take a look to add in more details and fix the mistakes.

File: src/transformers/generation/logits_process.py (hunk 7 above)
Comment (ArthurZucker) [reply src/transformers/generation/logits_process.py:1373]: My point is that the doc says it stops on poem.

File: src/transformers/generation/logits_process.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jprivera44) [reply src/transformers/generation/logits_process.py]: @ArthurZucker I've updated the docstring, please let me know what you think? I appreciate the help on this.

File: src/transformers/generation/logits_process.py (hunk 7 above)
Comment (jprivera44) [reply src/transformers/generation/logits_process.py]: Oh I see what you mean, and I just went through and made the fix.

File: src/transformers/generation/logits_process.py [hunk 9]
```diff
@@ -1330,8 +1330,12 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
 
 class WhisperTimeStampLogitsProcessor(LogitsProcessor):
     r"""
-    Whisper specific Processor. This processor can be used to force a list of tokens. The processor will set their log
-    probs to `inf` so that they are sampled at their corresponding index.
+
+    [`LogitsProcessor`] that modifies the logits for the generation of timestamps in the transcription. When the input
+    tokens are at a specific threshold, the processor sets the scores to negative infinity and returned. The timestamps
+    are processed to be in pairs except right before the end-of-sequence token. If the total probability of all
+    timestamp tokens is greater than any individual non-timestamp token, the processor sets those non-timestamp logits
```
Comment (ArthurZucker) [inline src/transformers/generation/logits_process.py:1337]:
```suggestion
    [`LogitsProcessor`] that modifies the logits for the generation of timestamps in the transcription. When the input
    tokens are at a specific threshold, the processor sets the scores to negative infinity. The processor makes sure that timestamp tokens appear in pair, by doing .... This is done because ......
     It also ensure that when  the predicted probability of sampling any of the timestamp token is greater than ........ This is done to ensure that..... 
```

**Review by ArthurZucker — COMMENTED** [review]
Hey! Again, let's make this helpful for users 😉

**Comment by jprivera44** [comment]
Hi sorry this got to the stale status, @ArthurZucker got in an update on this! Thank you again.

**Review by ArthurZucker — APPROVED** [review]
A lot better thanks 😉

**Comment by jprivera44** [comment]
No thank you guys! Apologies this took a while!
